apps/app_voicemail.c

```diff
@@ -1322,6 +1331,136 @@ static char *strip_control_and_high(const char *input, char *buf, size_t buflen)
 	return buf;
 }
 
+/*
+ * These macros are currently only used by the ODBC code but
+ * should be used anyplace we're allocating a PATH_MAX
+ * sized buffer for something which will be much smaller
+ * in practice.
+ */
+#ifdef ODBC_STORAGE
+/*!
+ * \internal
+ * \brief Get the length needed for a message path base.
+ *
+ * \param dir The absolute path to a directory
+ *
+ * \note The directory provided should include all componenets
+ * down to the folder level.  For example:
+ *  /var/spool/asterisk/voicemail/default/1234/INBOX
+ *
+ * \returns Number of bytes needed for the message path base.
+ */
+static size_t get_msg_path_len(const char *dir)
+{
+	if (ast_strlen_zero(dir)) {
+		return 0;
+	}
+	/* dir + '/' + MSGFILE_LEN + '\0' */
+	return strlen(dir) + 1 + MSGFILE_LEN + 1;
+}
+
+/*
+ * The longest sound file extension we currently
+ * have is "siren14" which is 7 characters long
+ * but we'll make this 12 to be extra safe.
+ */
+#define MAX_SOUND_EXTEN_LEN 12
+/*!
+ * \internal
+ * \brief Get the length needed for a message sound or metadata file path.
+ *
+ * \param dir The absolute path to a directory
+ *
+ * \note The directory provided should include all componenets
+ * down to the folder level.  For example:
+ *  /var/spool/asterisk/voicemail/default/1234/INBOX
+ *
+ * \returns Number of bytes needed for the message sound or metadata file path.
+ */
+static size_t get_msg_path_ext_len(const char *dir)
+{
+	if (ast_strlen_zero(dir)) {
+		return 0;
+	}
+	/* dir + '/' + MSGFILE_LEN + extlen(including '.') + '\0' */
+	return strlen(dir) + 1 + MSGFILE_LEN + MAX_SOUND_EXTEN_LEN + 1;
+}
+
+/*!
+ * \internal
+ * \brief Allocate a buffer on the stack containing a message file path base.
+ *
+ * \param dir The absolute path to a directory
+ * \param msgnum The message number or -1 for a prompt file
+ *
+ * \note The returned buffer will automatically be freed when it
+ * goes out of scope.
+ *
+ * If passed (/var/spool/asterisk/voicemail/default/1234/INBOX, 0)
+ * the returned path would be:
+ * /var/spool/asterisk/voicemail/default/1234/INBOX/msg0000
+ *
+ * If passed (/var/spool/asterisk/voicemail/default/1234/unavail, -1)
+ * the returned path would be:
+ * /var/spool/asterisk/voicemail/default/1234/unavail
+ * (no change)
+ *
+ * \returns A pointer to the buffer containing the path.
+ */
+#define MAKE_FILE_PTRA(dir, msgnum) \
+({ \
+    size_t __len = get_msg_path_len(dir); \
+    char *__var; \
+    if (msgnum < 0) { \
+        __var = ast_strdupa(dir); \
+    } else { \
+        __var = ast_alloca(__len); \
```

Comment:
Would be nice if there was an asprintf version that used alloca instead of alloc... do you think it's worth adding a macro for that to simplify all these?

Reply:
sprintfa?  That would get very confusing but I do like the idea.

Reply:
I'll see if I can clean these up.

Reply:
> So here's the deal... part of the convenience of the macros is using alloca so you don't have to call ast_free or use RAII_VAR. You can't wrap alloca in a function because of its very nature of allocating on the stack. You also can't use VA_ARGS because the only way to know how much memory to allocate would be to call snprintf once with a `0` as the buffer length to get the buffer size needed, then allocate the memory and call snprintf again. That's just a waste for these limited cases. I'm going to leave the macros as is.

I know a function definitely won't work, but won't a macro wrapper work?

Reply:
#define sprintf_alloca(fmt, ...) {( int l = snprintf(NULL, 0, fmt, ##__VA_ARGS__); char *b = alloca(l+1); if (b) { snprintf(b, l+1, fmt, ##__VA_ARGS__); }; b; })

What I don't like about this is the double-invocation of snprintf, which if __VA_ARGS__ contains function calls which isn't marked as pure could cause problems.

Reply:
alternatively asprintf(), strdupa(), free(), but that just feels like a waste of CPU cycles, not to mention abusing the memory manager even more (as if asterisk doesn't already cause malloc enough headaches).

Reply:
> #define sprintf_alloca(fmt, ...) {( int l = snprintf(NULL, 0, fmt, ##**VA_ARGS**); char *b = alloca(l+1); if (b) { snprintf(b, l+1, fmt, ##**VA_ARGS**); }; b; })

The second snprintf is not necessary, sprintf could be used safely in that case.

> What I don't like about this is the double-invocation of snprintf, which if **VA_ARGS** contains function calls which isn't marked as pure could cause problems.

But this is true in existing usage as well, no?

The macro could have a `note` that notes that.

> alternatively asprintf(), strdupa(), free(), but that just feels like a waste of CPU cycles, not to mention abusing the memory manager even more (as if asterisk doesn't already cause malloc enough headaches).

I agree, we should only add a macro if that will simplify without adding actual overhead, which I think is possible but I didn't think about it extensively.

Regarding length, I noticed this in a few places. The place I commented seems to already have the length so the benefit is less here, but in other places, the length is calculated for alloca in which case it seems like a no brainer to use a macro like this.

Reply:
perhaps two variants?  One that takes a length snprintf_alloca(len, fmt, ...) and sprintf_alloca(fmt, ...) where the latter uses snprintf to get the length and then just reverts to snprintf_alloca, being sure to *document* the double evaluation of arguments.  Just be wary of the off by one, so snprintf_alloca(len, fmt, ...) should alloca(len+1) such that len is the length of the string without the NULL (from printf man page "Upon successful return, these functions return the number of characters printed (excluding the null byte used to end output to strings)." so IMHO it makes sense to stick with the "excluding null byte" assumptions).

Reply:
So, it this just a philosophical discussion now or are you guys requesting changes?

Reply:
I would like to see a change, at least in the alloca case where it simplifies with no downside

Reply:
I believe @gtjoseph already eliminated the macro here, so we should probably move the specific discussion elsewhere.

Reply:
> I would like to see a change, at least in the alloca case where it simplifies with no downside

What change specifically?  I'm seeing neither complexity nor downside in the macros as is.

Reply:
Okay, I just reviewed again and now nothing rubs me the wrong way anymore so I think you addressed them all :)